Small commit that simply corrects the types in the coupon.ex to match…

kcraybould opened the pull request:
… the standards of the library.

This is in support of issues number 266 (https://github.com/code-corps/stripity_stripe/issues/266).  It would complete the subtask Subscriptions->Coupons->Clean Up Types.

joshsmith commented on lib/stripe/subscriptions/coupon.ex:
```diff
@@ -19,18 +19,18 @@ defmodule Stripe.Coupon do
+               max_redemptions: pos_integer | nil,
                metadata: %{
                  optional(String.t) => String.t
                },
```
Should change to `metadata: Stripe.Types.metadata`.

kcraybould replied:
This doesn't match the type definitions in the issue: metadata becomes %{ optional(String.t) => String.t }  Which should we go with?

joshsmith commented on lib/stripe/subscriptions/coupon.ex:
```diff
@@ -19,18 +19,18 @@ defmodule Stripe.Coupon do
-               currency: String.t,
-               duration: :forever | :once | :repeating,
+               amount_off: pos_integer | nil,
+               created: pos_integer,
```
Timestamps should be a `Stripe.timestamp()` with `| nil` if necessary (or any other additional constraints).

kcraybould replied:
The created was listed as an integer in the yaml and described as the number of seconds from the epoch.  Shouldn't those be kept as integers, or are we saying that we convert epoch values to timestamps?

joshsmith commented on lib/stripe/subscriptions/coupon.ex:
```diff
@@ -57,19 +55,19 @@ defmodule Stripe.Coupon do
         when params: %{
-               id: String.t,
+               id: String.t(),
                duration: :forever | :once | :repeating,
```
Duration here should be `String.t()`, I think.

joshsmith commented on lib/stripe/subscriptions/coupon.ex:
```diff
@@ -57,19 +55,19 @@ defmodule Stripe.Coupon do
                max_redemptions: pos_integer,
                metadata: %{
-                 optional(String.t) => String.t
+                 optional(String.t()) => String.t()
```
Should be `Stripe.Types.metadata()`.

joshsmith commented on lib/stripe/subscriptions/coupon.ex:
```diff
@@ -57,19 +55,19 @@ defmodule Stripe.Coupon do
                },
                percent_off: pos_integer,
-               redeem_by: Stripe.timestamp
+               redeem_by: Stripe.timestamp()
```
My guess is that a number of the `when params` here are actually optional. You can look at some of my recent PRs to see how I've handled that.

joshsmith commented on lib/stripe/subscriptions/coupon.ex:
```diff
@@ -124,12 +118,12 @@ defmodule Stripe.Coupon do
+               created: Stripe.date_query() | nil,
+               ending_before: t | Stripe.id() | nil,
+               limit: 1..100 | nil,
+               starting_after: t | Stripe.id() | nil
```
I don't think the `| nil` is the right approach here. Look to the instances where we use `optional`, for example in `Stripe.Charge`.

joshsmith commented:
@kcraybould I generally recommend not merging the branch into your branch, but instead doing a rebase onto the branch. I have some guides on how to do this if you're interested!

kcraybould commented:
@joshsmith Yes, whatever the preferred method here is I will follow.  At this stage, what do you recommend to bring this inline with the project standards?

kcraybould commented:
@begedin  you are right. There was one instance of a timestamp, redeem_by, that I had changed into a pos_integer   I corrected that.  The other instances of pos_integer aren't timestamps, so I assume they should remain pos_integers?

joshsmith commented:
@begedin we aren't abandoning the abstraction, just to clarify.

kcraybould commented:
All, after talking with @joshsmith  I am going to close this PR.  I managed to get into rebase hell with badly done merges, so I am going to save off the work, close this PR and fix the mess in a clean setup.  Sorry for the confusion.
